### PR TITLE
[`bnb`] Let's make serialization of int8 models possible

### PR DESCRIPTION
**Pull request by younesbelkada**
# What does this PR do?

Before this PR, it was not possible to save an 8bit model, or load an 8bit model from the Hub. This PR makes this feature possible. If this PR gets merged, users can upload 8bit models on the Hub and/or load 8bit models from the Hub, hence save 2x memory compared to half-precision models.

```python
from transformers import AutoModelForCausalLM, AutoTokenizer

model = AutoModelForCausalLM.from_pretrained("bigscience/bloom-560m", device_map="auto", load_in_8bit=True)
tokenizer = AutoTokenizer.from_pretrained("bigscience/bloom-560m")

text = "Hello my name is"
inputs = tokenizer(text, return_tensors="pt").to(0)
outputs = model.generate(**inputs)
print(tokenizer.decode(outputs[0]))

>>> Hello my name is Nate, I am a professional photographer and I am a member of the
model.save_pretrained("./saved_int8")

model = AutoModelForCausalLM.from_pretrained("./saved_int8")

outputs = model.generate(**inputs)
print(tokenizer.decode(outputs[0]))
>>> Hello my name is Nate, I am a professional photographer and I am a member of the
```

Depends on https://github.com/TimDettmers/bitsandbytes/pull/159 

Let's put it as draft before I address the last TODOs and open questions & before https://github.com/TimDettmers/bitsandbytes/pull/159  gets merged.

## TODOs and open questions:

- ability to push `BitsAndBytesConfig`
- Do we want to save the serialized model under the name `pytorch_model.bin` ? I would say yes for simplicity reasons but we need to make sure that a user calls `from_pretrained` with `load_in_8bit`, hence add a warning if there is a `quantization_config.json` on the Hub repo + the user is not passing `load_in_8bit=True`.
- Force `load_in_8bit=True` if there is a `quantization_config.json` on the Hub repo?
- Update docs
- Update warnings
- Safety checkers for `bnb` versions
- Add a test to check if it works using sharded fp16 weights

cc @sgugger I left few open questions, would love to hear your thoughts on these!

### CODE REVIEW

**Comment by sgugger** [comment]
The design is not easy enough to use. If a user saves a quantized model and pushes to the Hub, it should work directly with `from_pretrained`. This is why I insisted that the quantization config should be saved inside the model config. This way you won't need to have the user pass `load_in_8_bit=True`, as you can read it from the config.

**Comment by younesbelkada** [comment]
awesome ok, I'll work on that, so if there is a quantized config on the repo we should force-use `device_map=auto` & `load_in_8bit` in this case

File: src/transformers/modeling_utils.py [hunk 1]
```diff
@@ -2085,6 +2128,39 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
         variant = kwargs.pop("variant", None)
         use_safetensors = kwargs.pop("use_safetensors", None if is_safetensors_available() else False)
 
+        if is_bitsandbytes_available():
+            # TODO: uncomment this after the next release of bitsandbytes
+            # can_serialize_bnb = version.parse(
+            #     importlib_metadata.version("bitsandbytes")
+            # ) >= version.parse("0.37.2")
```
Comment (younesbelkada) [inline src/transformers/modeling_utils.py:2135]: This will be removed on the next `bitsandbytes` release

**Comment by younesbelkada** [comment]
The PR is ready for review @sgugger ! 
This PR is not mergeable before the bnb release of course

File: docs/source/en/main_classes/quantization.mdx [hunk 2]
```diff
@@ -52,6 +52,37 @@ Note that once a model has been loaded in 8-bit it is currently not possible to
...
+
+<Tip warning={true}>
+
+Pushing 8bit models on the Hub is strongely encouraged for large models. This will allow the community to benefit from the memory footprint reduction and loading for example large models on a Google Colab.
+
+</Tip>
+
+### Load a quantized model from the 🤗 Hub
+
+You can load a quantized model from the Hub by using `from_pretrained` method. Make sure that the pushed weights are quantized, by checking that the file `quantization_config.json` is present in the model repository.
```
Comment (sgugger) [inline docs/source/en/main_classes/quantization.mdx:77]: No it should be part of the model config, it doesn't need to be in its own file.

**Review by sgugger — COMMENTED** [review]
I think we didn't understand each other ;-) 
The quantization config should be saved as part of the model config (the model weights cannot be used without it anyway, once the model is quantized it's really part of the information necessary to load it) It will make all the download code irrelevant and this PR muche easier.

**Comment by younesbelkada** [comment]
Thanks for the heads up! :D 
It should be much better now! For me the PR is ready for a review now

File: src/transformers/utils/quantization_config.py [hunk 3]
```diff
@@ -14,11 +14,20 @@
...
+
+from ..utils import logging
+
+
+logger = logging.get_logger(__name__)
 
 
 @dataclass
-class BitsAndBytesConfig:
+class BitsAndBytesConfig(object):
```
Comment (sgugger) [inline src/transformers/utils/quantization_config.py:30]:
```suggestion
class BitsAndBytesConfig:
```
No need to add this.

**Review by sgugger — APPROVED** [review]
Nice, thanks a lot for iterating!

File: src/transformers/modeling_utils.py [hunk 4]
```diff
@@ -1673,10 +1681,10 @@ def save_pretrained(
                 Additional key word arguments passed along to the [`~utils.PushToHubMixin.push_to_hub`] method.
         """
         # Checks if the model has been loaded in 8-bit
-        if getattr(self, "is_loaded_in_8bit", False):
+        if getattr(self, "is_loaded_in_8bit", False) and getattr(self, "is_8bit_serializable", False):
             warnings.warn(
                 "You are calling `save_pretrained` to a 8-bit converted model you may likely encounter unexepected"
-                " behaviors. ",
+                " behaviors. If you want to save 8-bit models, make sure to have `bitsandbytes>=0.37.2` installed.",
```
Comment (younesbelkada) [inline src/transformers/modeling_utils.py:1687]:
```suggestion
                " behaviors. If you want to save 8-bit models, make sure to have `bitsandbytes>0.37.2` installed.",
```

File: docs/source/en/main_classes/quantization.mdx [hunk 5]
```diff
@@ -52,6 +52,37 @@ Note that once a model has been loaded in 8-bit it is currently not possible to
...
+</Tip>
+
+### Load a quantized model from the 🤗 Hub
+
+You can load a quantized model from the Hub by using `from_pretrained` method. Make sure that the pushed weights are quantized, by checking that the attribute `quantization_config` is present in the model configuration object.
+
+```python
+from transformers import AutoModelForCausalLM, AutoTokenizer
+
+model = AutoModelForCausalLM.from_pretrained("ybelkada/bloom-560m-8bit")
```
Comment (sgugger) [inline docs/source/en/main_classes/quantization.mdx:82]:
```suggestion
model = AutoModelForCausalLM.from_pretrained("{your_username}/bloom-560m-8bit")
```

File: src/transformers/utils/bitsandbytes.py [hunk 6]
```diff
@@ -61,14 +65,22 @@ class `Int8Params` from `bitsandbytes`.
             elif isinstance(value, torch.Tensor):
                 new_value = value.to("cpu")
                 if value.dtype == torch.int8:
-                    raise ValueError(
-                        "You cannot load weights that are saved in int8 using `load_in_8bit=True`, make sure you are",
-                        " using `load_in_8bit=True` on float32/float16/bfloat16 weights.",
+                    # TODO: @younesbelkada check bnb version
```
Comment (sgugger) [inline src/transformers/utils/bitsandbytes.py:68]:
```suggestion
```
Seems like this should be removed.

**Review by sgugger — APPROVED** [review]
LGTM with a couple last nits!

**Pull request merged**